modules/lang-painless/src/main/java/org/elasticsearch/painless/Def.java

```diff
@@ -1286,21 +1279,154 @@ private ArrayIndexNormalizeHelper() {}
         public final int numCaptures;
         public final String encoding;
 
+        private static final String FORMAT = "[SD][tf]symbol.methodName,numCaptures";
+
         public Encoding(boolean isStatic, boolean needsInstance, String symbol, String methodName, int numCaptures) {
             this.isStatic = isStatic;
             this.needsInstance = needsInstance;
-            this.symbol = symbol;
-            this.methodName = methodName;
+            this.symbol = Objects.requireNonNull(symbol);
+            this.methodName = Objects.requireNonNull(methodName);
+            this.numCaptures = numCaptures;
+            this.encoding = encoding();
```

Comment:
Is there any chance we could build the encoding on demand? I think it's probably nice to avoid the extra sting building stuff most of the time. We only will need it for debugging and validation, right?

Reply:
If we're using this constructor then we're going to need the encoding.  The encoding becomes `recipeString` in [`Def.lookupMethod`](https://github.com/elastic/elasticsearch/blob/master/modules/lang-painless/src/main/java/org/elasticsearch/painless/Def.java#L188).

`invokedynamic` can only take [certain types] of static arguments, one of which is `String`, so we can't use our lovely custom type there.

> If the call site specifier has any static arguments, then a parameter type for each argument is appended to the parameter types of the method descriptor in the order that the arguments were pushed on to the operand stack. These parameter types may be Class, java.lang.invoke.MethodHandle, java.lang.invoke.MethodType, String, int, long, float, or double.

Reply:
Ah well. I think I remember us always using the string when I reviewed it the first time. Maybe worth a comment, but otherwise, :+1:

Reply:
Added a comment.